_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+
+    return (
+      <div className={`Drawer ${drawerStyles}`}>
+        <div
+          className="Drawer-title"
+          onClick={this.handleTitleClick}
+        >
+          {title}
+        </div>
+        <div className="Drawer-content">
+          {children}
+        </div>
+      </div>
+    );
+  }
+}
+
+export default Drawer;
+
+/////////////////////////////////////
+// And here's how you would use it //
+/////////////////////////////////////
+import React from 'react';
+import Drawer from '~/components/Drawer'; // The location of our new component
+
+class AnotherComponent extends React.Component {
+  // code omitted
+  render() {
+    return (
+      <div className="AnotherComponent">
+        <h1>Welcome to my component</h1>
+        <Drawer title="My Drawer Title">
+          <ul>
+            <li>Item 1</li>
+            <li>Item 2</li>
+            <li>Item 3</li>
+          </ul>
+        </Drawer>
+      </div>
+    );
+  }
+}
+```
+
+Here's what it looks like so far. Really the only thing it does is provide
+structure and toggle a CSS class `is-expanded` based on state.
+
+![Unstyled Drawer Component](/img/2019/unstyled-drawer-component.gif)
+
+It's pretty ugly at this point, so in the next section we're going to add some
+SCSS styles to it.
+
+## Adding Style
+
+One thing missing from our component at the moment is that the Drawer's content
+is always shown. We only want it to display if the state of the component is expanded. We're going to eventually animate it into existing using a slide transition so focusing on the element's height will be advantageous here.
+
+Start off by hiding the content using the following style
+configuration.
+
+``` scss
+.Drawer {
+  &-content {
+    height: 0;
+    overflow: hidden;
+  }
+}
+```
+
+We need both `height: 0` and `overflow: hidden` here as they work in tandem. Having a
+height of 0 still shows the content even with a container without height. Adding overflow hidden ensures that the content in the container outside of the height will remain hidden. These two together remove the content.
+
+Next we probably want the `Drawer-title` element to look like it is actually clickable. We can do this by adding a border, background, and cursor styles below.
+
+``` scss
+.Drawer {
+  &-title {
+    background: #efefef;
+    border: 1px solid;
+    cursor: pointer;
+    display: inline-block;
+    padding: 16px;
+  }
+
+  &-content {
+    ...
+  }
+}
+```
+
+The result is a primitive looking button to click. Which brings up something else we should probably fix at this point which is keyboard accessibility. Time for a slight detour.
+
+![Basic title styles](/img/2019/basic-title-style.png)
+
+### Keyboard Accessibility
+
+Right now on the page anyone using a screen reader or other non-mouse assistance technology, will be unable to tab to our new button. One way we could attempt to fix this is by giving our `Drawer-title` div a `tabIndex` value and a `role` value.
+
```

Comment:
One thing I recently learned about was accessibility of websites and applications. For example, anyone using a screen reader or other non-mouse assistance technology will be unable to use to our new button. This is terrible because it's a cool button and we want to be sure we develop with everyone in mind. (Check out this link for more on accessibility). One way we could attempt to fix this is by giving our `Drawer-title` div a `tabIndex` value and a `role` value.